core/capabilities/ccip/ccipaptos/msghasher.go

```diff
@@ -0,0 +1,367 @@
+		if err != nil {
+			return [32]byte{}, fmt.Errorf("decode dest gas amount: %w", err)
+		}
+
+		lggr.Debugw("decoded dest gas amount",
+			"destGasAmount", destGasAmount)
+
+		destTokenAddress, err := addressBytesToBytes32(rta.DestTokenAddress)
+		if err != nil {
+			return [32]byte{}, fmt.Errorf("decode dest token address: %w", err)
+		}
+
+		lggr.Debugw("abi decoded dest token address",
+			"destTokenAddress", destTokenAddress)
+
+		rampTokenAmounts = append(rampTokenAmounts, any2AptosTokenTransfer{
+			SourcePoolAddress: rta.SourcePoolAddress,
+			DestTokenAddress:  destTokenAddress,
+			DestGasAmount:     destGasAmount,
+			ExtraData:         rta.ExtraData,
+			Amount:            rta.Amount.Int,
+		})
+	}
+
+	// one difference from EVM is that we don't left pad the OnRamp to 32 bytes here, we use the source chain's canonical bytes encoding directly.
+	metaDataHashInput, err := computeMetadataHash(uint64(msg.Header.SourceChainSelector), uint64(msg.Header.DestChainSelector), msg.Header.OnRamp)
+	if err != nil {
+		return [32]byte{}, fmt.Errorf("abi encode metadata hash input: %w", err)
+	}
+
+	lggr.Debugw("metadata hash preimage",
+		"metaDataHashInput", hexutil.Encode(metaDataHashInput[:]))
+
+	// Need to decode the extra args to get the gas limit.
+	// TODO: we assume that extra args is always abi-encoded for now, but we need
+	// to decode according to source chain selector family. We should add a family
+	// lookup API to the chain-selectors library.
+
+	decodedExtraArgsMap, err := h.extraDataCodec.DecodeExtraArgs(msg.ExtraArgs, msg.Header.SourceChainSelector)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	gasLimit, err := parseExtraDataMap(decodedExtraArgsMap)
+	if err != nil {
+		return [32]byte{}, fmt.Errorf("decode extra args to get gas limit: %w", err)
+	}
+
+	lggr.Debugw("decoded msg gas limit", "gasLimit", gasLimit)
+
+	receiverAddress, err := addressBytesToBytes32(msg.Receiver)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	msgHash, err := computeMessageDataHash(metaDataHashInput, msg.Header.MessageID, receiverAddress, uint64(msg.Header.SequenceNumber), gasLimit, msg.Header.Nonce, msg.Sender, msg.Data, rampTokenAmounts)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	lggr.Debugw("final message hash result",
+		"msgHash", hexutil.Encode(msgHash[:]),
+	)
+
+	return msgHash, nil
+}
+
+// This is the equivalent of offramp::calculate_message_hash.
+// This is very similar to the EVM version, except for 32-byte addresses:
+//
+//	bytes32 messageDataHash = keccak256(
+//	  abi.encode(
+//	    LEAF_DOMAIN_SEPARATOR,
+//	    metadataHash,
+//	    keccak256(
+//	      abi.encode(
+//	        original.header.originalId,
+//	        original.receiver,
+//	        original.header.sequenceNumber,
+//	        original.gasLimit,
+//	        original.header.nonce
+//	      )
+//	    ),
+//	    keccak256(original.sender),
+//	    keccak256(original.data),
+//	    keccak256(abi.encode(original.tokenAmounts))
+//	  )
+//	);
```

Comment:
Lets paste the move code that does the hashing here to be more precise, and make it easier to review.